### DISCUSSION TITLE
Questions about upgrading vuepress project

### DISCUSSION THREAD

**Discussion by zwik** [vuepress/core#1197] [Q&A] [answered]
### Description

When upgrading I can't use a dev nor production build and get error during development like this:
```
[plugin:vite:import-analysis] Failed to resolve import "/home/marcel/documents/code/environment/mirror-2021/node_modules/@vuepress/theme-default/lib/client/layouts/404.vue" from "docs/.vuepress/.temp/internal/layoutComponents.js". Does the file exist?
```

I removed the caret and installed again but same issues.

### Reproduction

https://github.com/zwik/mirror-2021

### Used Package Manager

npm

### System Info

```shell
npm run info                                                                                                                                                                                    127 ✘  09:31:47  

> mirror-2021@1.0.0 info
> vuepress info


  System:
    OS: Linux 5.15 Ubuntu 22.10 22.10 (Kinetic Kudu)
    CPU: (12) x64 AMD Ryzen 5 3600 6-Core Processor
    Memory: 12.66 GB / 15.58 GB
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 1.22.5 - /mnt/c/Program Files (x86)/Yarn/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
  Utilities:
    Git: 2.37.2 - /usr/bin/git
  Browsers:
    Chrome: Not Found
    Firefox: Not Found
  npmPackages:
    @vuepress/bundler-vite:  2.0.0-beta.58 
    @vuepress/bundler-webpack: Not Found
    @vuepress/cli:  2.0.0-beta.58 
    @vuepress/client:  2.0.0-beta.50-pre.1 (2.0.0-beta.58)
    @vuepress/core:  2.0.0-beta.58 
    @vuepress/markdown:  2.0.0-beta.58 
    @vuepress/plugin-active-header-links:  2.0.0-beta.58 
    @vuepress/plugin-back-to-top:  2.0.0-beta.58 
    @vuepress/plugin-container:  2.0.0-beta.58 
    @vuepress/plugin-docsearch: Not Found
    @vuepress/plugin-external-link-icon:  2.0.0-beta.58 
    @vuepress/plugin-git:  2.0.0-beta.58 
    @vuepress/plugin-google-analytics: Not Found
    @vuepress/plugin-medium-zoom:  2.0.0-beta.58 
    @vuepress/plugin-nprogress:  2.0.0-beta.58 
    @vuepress/plugin-palette:  2.0.0-beta.58 
    @vuepress/plugin-prismjs:  2.0.0-beta.58 
    @vuepress/plugin-pwa: Not Found
    @vuepress/plugin-pwa-popup: Not Found
    @vuepress/plugin-register-components: Not Found
    @vuepress/plugin-search: Not Found
    @vuepress/plugin-shiki: Not Found
    @vuepress/plugin-theme-data:  2.0.0-beta.58 
    @vuepress/plugin-toc: Not Found
    @vuepress/shared:  2.0.0-beta.50-pre.1 (2.0.0-beta.58)
    @vuepress/theme-default:  2.0.0-beta.58 
    @vuepress/utils:  2.0.0-beta.58 
    vue:  3.2.45 
    vue-loader: Not Found
    vue-router:  4.1.6 
    vuepress: 2.0.0-beta.58 => 2.0.0-beta.58 
    vuepress-vite:  2.0.0-beta.58 
    vuepress-webpack: Not Found
```

**Comment by init-qy** [comment, marked as answer]
see #1195, don't use `^`

**Comment by zwik** [comment]
As mentioned in the description. I've tried that, but that doesn't seem to work for me. I've updated the repo with that as well. You can also see in the output of `vuepress info` that the caret is not there.

**Comment by init-qy** [comment]
> As mentioned in the description. I've tried that, but that doesn't seem to work for me. I've updated the repo with that as well. You can also see in the output of `vuepress info` that the caret is not there.

remove the `package-lock.json`, and try again?

in the `vuepress info`, @vuepress/client:  2.0.0-beta.50-pre.1 (2.0.0-beta.58) looks incorrent

this is my `vuepress info`

```
  System:
    OS: Windows 10 10.0.19044
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 5.00 GB / 15.91 GB
  Binaries:
    Node: 16.13.1 - C:\Program Files\nodejs\node.EXE
    Yarn: Not Found
    npm: 8.1.2 - C:\Program Files\nodejs\npm.CMD
  Utilities:
    Git: 2.29.2.
  Browsers:
    Chrome: 108.0.5359.99
    Edge: Spartan (44.19041.1266.0), Chromium (108.0.1462.46)
  npmPackages:
    @vuepress/bundler-vite:  2.0.0-beta.58
    @vuepress/bundler-webpack: Not Found
    @vuepress/cli:  2.0.0-beta.58
    @vuepress/client: 2.0.0-beta.58 => 2.0.0-beta.58
    @vuepress/core:  2.0.0-beta.58
    @vuepress/markdown:  2.0.0-beta.58
    @vuepress/plugin-active-header-links:  2.0.0-beta.58
    @vuepress/plugin-back-to-top:  2.0.0-beta.58
    @vuepress/plugin-container:  2.0.0-beta.58
    @vuepress/plugin-docsearch: Not Found
    @vuepress/plugin-external-link-icon:  2.0.0-beta.58
    @vuepress/plugin-git:  2.0.0-beta.58
    @vuepress/plugin-google-analytics: Not Found
    @vuepress/plugin-medium-zoom:  2.0.0-beta.58
    @vuepress/plugin-nprogress:  2.0.0-beta.58
    @vuepress/plugin-palette:  2.0.0-beta.58
    @vuepress/plugin-prismjs:  2.0.0-beta.58
    @vuepress/plugin-pwa: Not Found
    @vuepress/plugin-pwa-popup: Not Found
    @vuepress/plugin-register-components: Not Found
    @vuepress/plugin-search: Not Found
    @vuepress/plugin-shiki: Not Found
    @vuepress/plugin-theme-data:  2.0.0-beta.58
    @vuepress/plugin-toc: Not Found
    @vuepress/shared:  2.0.0-beta.58
    @vuepress/theme-default:  2.0.0-beta.58
    @vuepress/utils:  2.0.0-beta.58
    vue: 3.2.45 => 3.2.45
    vue-loader: Not Found
    vue-router: 4.1.6 => 4.1.6
    vuepress: 2.0.0-beta.58 => 2.0.0-beta.58
    vuepress-vite:  2.0.0-beta.58
    vuepress-webpack: Not Found

```

**Comment by zwik** [comment]
I've just removed the `package-lock.json` (and removed node_modules) and did `npm install` again, but it generates the exact same lock file. It will also result in the same error.

**Comment by zwik** [comment]
I've installed `@vuepress/client@next` as well now and that seems to resolve the issue.

When I look at the documentation I see that `pnpm` docs do require to install that dependency but `yarn` and `npm` docs do not mention that. Is the documentation correct?

**Comment by init-qy** [comment]
I find the peerDependencies in `vuepress-vite`, maybe this cause the incorrent version of `@vuepress/client`
In my view, `pnpm` is different when deal with `peer`, so it needs to install `@vuepress/client@next`
https://github.com/vuepress/vuepress-next/blob/619fc129246183658afc673febb7002f41970276/ecosystem/vuepress-vite/package.json#L47-L50

**Comment by Mister-Hope** [comment]
> I've installed `@vuepress/client@next` as well now and that seems to resolve the issue.
> 
> When I look at the documentation I see that `pnpm` docs do require to install that dependency but `yarn` and `npm` docs do not mention that. Is the documentation correct?

Yes, when using npm, any package can import other packages in the project, while pnpm only allow packages to import packages which declared in their `package.json`. `@vuepress/client` and `vue` are used during your usage, so they are required.

**Comment by Mister-Hope** [comment]
This issue is converted to a dicussion as it's dumplicate of https://github.com/vuepress/vuepress-next/issues/1195, however you are still welcomed to ask question about upgrade issues or solution.

**Comment by zwik** [comment]
Since the issue is now closed I'm still wondering what should be improved. I understand that `pnpm` works differently but apparently the `peerDependencies` are not correct. Are those going to be updated?

**Reply by Mister-Hope** [reply]
Yes, we may update that by making a script.

 Cc @meteorlxy shall I open a issue to remind this? We may run a custom script while bumpp to upgrade our perr version

**Reply by Mister-Hope** [reply]
You are welcome to open an issue to remind us just in case we forget it.

**Reply by zwik** [reply]
Can't we use https://github.com/vuepress/vuepress-next/issues/1196 for that?

**Reply by meteorlxy** [reply]
The wrong peer-deps was caused by [an issue of pnpm](https://github.com/pnpm/pnpm/issues/4164). I just updated it with a workaround and published in beta.59. Maybe you can have a try.

**Reply by zwik** [reply]
I've upgraded Vuepress now to beta 59 and removed the @vuepress/client dependency again, and did an `npm ci` and that indeed does seem to resolve the issue. Thanks!
